### ISSUE TITLE
Consistency check between headers and documentation

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#13442]
It would be very nice if we could extend doc-nits to confirm consistency between the declaration of a function in the headers, and the declaration of a function in the documentation. This might prevent errors such as #13441 from creeping in.

**Comment by levitte** [comment]
That would be nice, but quite frankly, would it not be better with something that can generate it more from the headers?  We have mentioned documentation generators in diverse meetings.

Either way, though, it's certainly possible to do! ... with some work.  I'd suggest expanding OpenSSL::ParseC to catch that sort of thing, if needed.  We must also somehow deal with functions that are really macros, but where the synopsis is still in C function declaration for (which is legitimate, mind you).

**Comment by mattcaswell** [comment]
> generate it more from the headers

Well that would ensure the declarations in the code are the same as in the headers. But actually I think that just creates a different problem. If the headers and docs are out of sync the thing we can check easily is the declaration - but if the declaration is out of sync then almost certainly the description is out of sync too. And we want to prompt the dev to take a look at the description.

**Comment by paulidale** [comment]
Ideal would be generating the prototypes from the headers and flagging if the documentation doesn't reference all arguments.  A more state-full parser would be needed.

**Comment by DDvO** [comment]
> Ideal would be generating the prototypes from the headers

Yes. All items listed in the NAME section should be auto-filled in the SYNOPSIS.

> and flagging if the documentation doesn't reference all arguments. A more state-full parser would be needed.

That's actually a different topic. See #15434.

**Comment by richsalz** [comment]
> flagging if the documentation doesn't reference all arguments. 

This would be a bad mistake: it prevents people from writing things like `foo does the same thing as foo_ex except that libctx is NULL.`  That doesn't need a full-state parser, it needs a full English parser. :)

**Comment by DDvO** [comment]
Good point, false positives would easily come up here.
They could be avoided by some heuristics and explicit exclusion pragmas, but this can become awkward.

Yet as I just mentioned before, argument documentation deviates from the original topic of this FR, and any further discussion on it should go to #15434.

**Comment by nhorman** [comment]
marking as inactive, to be closed at the end of 3.4 dev, barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
